File: chatbot-service/src/backend/Novit.Web.Api/Services/NurturingMailService.cs
using System.Net;
using System.Net.Mail;
using System.Text.RegularExpressions;
using MailKit;
using MailKit.Net.Imap;
using MailKit.Search;
using Microsoft.Extensions.Options;
using MimeKit;
using Novit.Web.Api.Models;

namespace Novit.Web.Api.Services;

/// <summary>
/// IMAP/SMTP mail service for the nurturing flow.
/// Supports separate login credentials and sender (alias) address — e.g. authenticate as
/// nicolasp@novitsoftware.com and send emails as Nicolás Piccardo &lt;nicolasp@novitsoftware.com&gt;.
/// Works with Gmail / Google Workspace, Zoho, or any standard IMAP+SMTP provider.
/// </summary>
public sealed class NurturingMailService : INurturingMailService
{
    private readonly NurturingOptions _options;
    private readonly ILogger<NurturingMailService> _logger;

    public NurturingMailService(IOptions<NurturingOptions> options, ILogger<NurturingMailService> logger)
    {
        _options = options.Value;
        _logger = logger;
    }

    public bool IsConfigured => _options.IsConfigured;

    public async Task<IReadOnlyList<IncomingEmail>> GetUnansweredRepliesAsync(int days = 30, CancellationToken ct = default)
    {
        if (!IsConfigured)
        {
            _logger.LogWarning("Nurturing mail not configured — skipping unanswered reply check");
            return [];
        }

        var results = new List<IncomingEmail>();

        try
        {
            using var client = new ImapClient();
            await client.ConnectAsync(_options.ImapHost, _options.ImapPort, MailKit.Security.SecureSocketOptions.SslOnConnect, ct);
            await client.AuthenticateAsync(_options.EffectiveLoginEmail, _options.EmailPassword, ct);

            var inbox = client.Inbox;
            await inbox.OpenAsync(FolderAccess.ReadOnly, ct);

            var since = DateTimeOffset.UtcNow.AddDays(-days);

            // Exclude messages sent from *both* the login account and the sender alias
            var loginEmail = _options.EffectiveLoginEmail!;
            var senderEmail = _options.EffectiveFromAddress!;

            var query = SearchQuery.DeliveredAfter(since.DateTime)
                .And(SearchQuery.Not(SearchQuery.FromContains(loginEmail)));

            // If the sender alias is different from the login, also exclude it
            if (!string.Equals(loginEmail, senderEmail, StringComparison.OrdinalIgnoreCase))
            {
                query = query.And(SearchQuery.Not(SearchQuery.FromContains(senderEmail)));
            }

            var uids = await inbox.SearchAsync(query, ct);

            // ── Build set of Message-IDs that WE sent (from Sent folder) ──
            // This is used to (a) detect already-answered threads and (b) verify
            // that incoming emails are genuine replies to messages we originated.
            var sentMessageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var answeredMessageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

            var sentFolder = (await client.GetFoldersAsync(client.PersonalNamespaces[0], cancellationToken: ct))
                .FirstOrDefault(f => f.Attributes.HasFlag(FolderAttributes.Sent))
                ?? (await client.GetFoldersAsync(client.PersonalNamespaces[0], cancellationToken: ct))
                    .FirstOrDefault(f => f.Name.Contains("Sent", StringComparison.OrdinalIgnoreCase));

            if (sentFolder is not null)
            {
                await sentFolder.OpenAsync(FolderAccess.ReadOnly, ct);
                var sentUids = await sentFolder.SearchAsync(SearchQuery.DeliveredAfter(since.DateTime), ct);
                foreach (var uid in sentUids)
                {
                    var sentMsg = await sentFolder.GetMessageAsync(uid, ct);

                    // Track every message we sent so we can verify In-Reply-To
                    if (!string.IsNullOrEmpty(sentMsg.MessageId))
                        sentMessageIds.Add(sentMsg.MessageId);

                    // Track which threads we already replied to
                    if (!string.IsNullOrEmpty(sentMsg.InReplyTo))
                        answeredMessageIds.Add(sentMsg.InReplyTo);
                }
            }

            _logger.LogInformation("Found {SentCount} sent message IDs and {AnsweredCount} answered thread IDs",
                sentMessageIds.Count, answeredMessageIds.Count);

            // Re-open inbox — opening sentFolder above implicitly closed it
            // (IMAP only allows one folder open per session)
            await inbox.OpenAsync(FolderAccess.ReadOnly, ct);

            foreach (var uid in uids)
            {
                var message = await inbox.GetMessageAsync(uid, ct);
                var messageId = message.MessageId ?? "";

                // Skip if we already replied to this message
                if (answeredMessageIds.Contains(messageId))
                    continue;

                // ── CRITICAL SAFETY CHECK ──
                // Only process emails that are genuine replies to messages WE sent.
                // This prevents auto-replying to spam, vendor newsletters, Google
                // notifications, or any other email that just happens to be in the inbox.
                var inReplyTo = message.InReplyTo;
                var isReplyToOurMessage = !string.IsNullOrEmpty(inReplyTo) && sentMessageIds.Contains(inReplyTo);

                // Also check References header as a fallback (some clients put it there)
                if (!isReplyToOurMessage && message.References is { Count: > 0 })
                {
                    isReplyToOurMessage = message.References.Any(r => sentMessageIds.Contains(r));
                }

                if (!isReplyToOurMessage)
                {
                    _logger.LogDebug("Skipping email from {From} ({Subject}) — not a reply to our sent messages",
                        message.From.Mailboxes.FirstOrDefault()?.Address, message.Subject);
                    continue;
                }

                results.Add(new IncomingEmail(
                    MessageId: messageId,
                    From: message.From.Mailboxes.FirstOrDefault()?.Address ?? "",
                    Subject: message.Subject ?? "",
                    Body: message.TextBody ?? message.HtmlBody ?? "",
                    Date: message.Date,
                    InReplyTo: message.InReplyTo));
            }

            await client.DisconnectAsync(true, ct);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to fetch unanswered replies via IMAP");
        }

        return results;
    }

    public async Task SendEmailAsync(string to, string subject, string body, bool isHtml = false, CancellationToken ct = default)
    {
        if (!IsConfigured)
        {
            _logger.LogWarning("Nurturing email not configured — skipping send to {To}", to);
            return;
        }

        using var smtp = CreateSmtpClient();

        var msg = new MailMessage(
            new MailAddress(_options.EffectiveFromAddress!, _options.SenderName),
            new MailAddress(to))
        {
            Subject = subject
        };

        if (isHtml)
            AttachHtmlWithPlainTextAlternative(msg, body);
        else
        {
            msg.Body = body;
            msg.IsBodyHtml = false;
        }

        await smtp.SendMailAsync(msg, ct);
        _logger.LogInformation("Nurturing email sent to {To}: {Subject}", to, subject);

        await LabelLastSentMessageAsync(subject, ct);
    }

    public async Task SendBulkEmailAsync(IEnumerable<string> recipients, string subject, string body, bool isHtml = false, CancellationToken ct = default)
    {
        if (!IsConfigured) return;

        var recipientList = recipients.ToList();
        if (recipientList.Count == 0) return;

        using var smtp = CreateSmtpClient();
        var from = new MailAddress(_options.EffectiveFromAddress!, _options.SenderName);

        var msg = new MailMessage
        {
            From = from,
            Subject = subject
        };

        foreach (var recipient in recipientList)
            msg.To.Add(new MailAddress(recipient));

        if (isHtml)
            AttachHtmlWithPlainTextAlternative(msg, body);
        else
        {
            msg.Body = body;
            msg.IsBodyHtml = false;
        }

        await smtp.SendMailAsync(msg, ct);
        _logger.LogInformation("Nurturing email sent to {Recipients}: {Subject}",
            string.Join(", ", recipientList), subject);

        await LabelLastSentMessageAsync(subject, ct);
    }

    public async Task SendReplyAsync(string to, string subject, string body, string inReplyToMessageId, bool isHtml = false, CancellationToken ct = default)
    {
        if (!IsConfigured) return;

        using var smtp = CreateSmtpClient();

        // Normalize subject: Gmail threads by subject, so preserve the original without adding extra "Re:"
        var replySubject = subject.StartsWith("Re:", StringComparison.OrdinalIgnoreCase) ? subject : $"Re: {subject}";

        var msg = new MailMessage(
            new MailAddress(_options.EffectiveFromAddress!, _options.SenderName),
            new MailAddress(to))
        {
            Subject = replySubject
        };

        if (isHtml)
            AttachHtmlWithPlainTextAlternative(msg, body);
        else
        {
            msg.Body = body;
            msg.IsBodyHtml = false;
        }

        msg.Headers.Add("In-Reply-To", inReplyToMessageId);
        msg.Headers.Add("References", inReplyToMessageId);

        await smtp.SendMailAsync(msg, ct);
        _logger.LogInformation("Nurturing reply sent to {To}: {Subject}", to, replySubject);

        await LabelLastSentMessageAsync(replySubject, ct);
    }

    public async Task MarkAsReadAsync(IEnumerable<string> messageIds, CancellationToken ct = default)
    {
        if (!IsConfigured) return;

        var idsToMark = messageIds.Where(id => !string.IsNullOrWhiteSpace(id)).ToHashSet(StringComparer.OrdinalIgnoreCase);
        if (idsToMark.Count == 0) return;

        try
        {
            using var client = new ImapClient();
            await client.ConnectAsync(_options.ImapHost, _options.ImapPort, MailKit.Security.SecureSocketOptions.SslOnConnect, ct);
            await client.AuthenticateAsync(_options.EffectiveLoginEmail, _options.EmailPassword, ct);

            var nurturingFolder = await GetOrCreateNurturingFolderAsync(client, ct);

            var inbox = client.Inbox;
            await inbox.OpenAsync(FolderAccess.ReadWrite, ct);

            // Search recent messages and match by Message-ID header
            var since = DateTimeOffset.UtcNow.AddDays(-60);
            var uids = await inbox.SearchAsync(SearchQuery.DeliveredAfter(since.DateTime), ct);

            var marked = 0;
            var archived = 0;
            var labeled = 0;
            foreach (var uid in uids)
            {
                if (idsToMark.Count == 0) break;

                var headers = await inbox.GetHeadersAsync(uid, ct);
                var msgId = headers[HeaderId.MessageId];
                if (msgId is not null && idsToMark.Remove(msgId))
                {
                    await inbox.AddFlagsAsync(uid, MessageFlags.Seen, silent: true, ct);
                    marked++;

                    // Prefer moving processed replies out of Inbox entirely so the mailbox
                    // owner no longer sees them once the bot consumed them.
                    if (nurturingFolder is not null)
                    {
                        try
                        {
                            await inbox.MoveToAsync(uid, nurturingFolder, ct);
                            archived++;
                            continue;
                        }
                        catch (Exception ex)
                        {
                            _logger.LogWarning(ex, "Could not archive inbox message into Nurturing; falling back to labeling");
                        }

                        try
                        {
                            await inbox.CopyToAsync(uid, nurturingFolder, ct);
                            labeled++;
                        }
                        catch (Exception ex)
                        {
                            _logger.LogWarning(ex, "Could not label inbox message as Nurturing");
                        }
                    }
                }
            }

            await client.DisconnectAsync(true, ct);
            _logger.LogInformation("Marked {Marked} emails as read, archived {Archived}, labeled {Labeled} as Nurturing", marked, archived, labeled);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to mark emails as read via IMAP");
        }
    }

    // ── IMAP Nurturing label helpers ──────────────────────────────

    private const string NurturingLabelName = "Nurturing";

    /// <summary>
    /// Gets or creates the "Nurturing" IMAP folder/Gmail label.
    /// Returns null if creation fails (non-fatal).
    /// </summary>
    private async Task<IMailFolder?> GetOrCreateNurturingFolderAsync(ImapClient client, CancellationToken ct)
    {
        try
        {
            var personal = client.PersonalNamespaces[0];
            var folders = await client.GetFoldersAsync(personal, cancellationToken: ct);
            var existing = folders.FirstOrDefault(f =>
                f.Name.Equals(NurturingLabelName, StringComparison.OrdinalIgnoreCase));

            if (existing is not null)
                return existing;

            // Create the label — Gmail exposes labels as top-level IMAP folders
            var topLevel = client.GetFolder(personal);
            var created = await topLevel.CreateAsync(NurturingLabelName, isMessageFolder: true, ct);
            _logger.LogInformation("Created Gmail label '{Label}'", NurturingLabelName);
            return created;
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Could not get/create '{Label}' IMAP folder", NurturingLabelName);
            return null;
        }
    }

    /// <summary>
    /// After sending an email via SMTP, finds the most recent message in the Sent folder
    /// matching the subject and copies it to the "Nurturing" label.
    /// Gmail IMAP COPY adds the label without duplicating the message.
    /// </summary>
    private async Task LabelLastSentMessageAsync(string subject, CancellationToken ct)
    {
        try
        {
            using var client = new ImapClient();
            await client.ConnectAsync(_options.ImapHost, _options.ImapPort, MailKit.Security.SecureSocketOptions.SslOnConnect, ct);
            await client.AuthenticateAsync(_options.EffectiveLoginEmail, _options.EmailPassword, ct);

            var nurturingFolder = await GetOrCreateNurturingFolderAsync(client, ct);
            if (nurturingFolder is null)
            {
                await client.DisconnectAsync(true, ct);
                return;
            }

            // Find the Sent folder
            var sentFolder = (await client.GetFoldersAsync(client.PersonalNamespaces[0], cancellationToken: ct))
                .FirstOrDefault(f => f.Attributes.HasFlag(FolderAttributes.Sent))
                ?? (await client.GetFoldersAsync(client.PersonalNamespaces[0], cancellationToken: ct))
                    .FirstOrDefault(f => f.Name.Contains("Sent", StringComparison.OrdinalIgnoreCase));

            if (sentFolder is null)
            {
                _logger.LogWarning("Could not find Sent folder to label message");
                await client.DisconnectAsync(true, ct);
                return;
            }

            await sentFolder.OpenAsync(FolderAccess.ReadOnly, ct);

            // Search for messages sent in the last 5 minutes with matching subject
            var since = DateTimeOffset.UtcNow.AddMinutes(-5);
            var query = SearchQuery.DeliveredAfter(since.DateTime)
                .And(SearchQuery.SubjectContains(subject.Length > 60 ? subject[..60] : subject));
            var uids = await sentFolder.SearchAsync(query, ct);

            if (uids.Count > 0)
            {
                // Copy the most recent match to the Nurturing label
                var lastUid = uids[^1];
                await sentFolder.CopyToAsync(lastUid, nurturingFolder, ct);
                _logger.LogDebug("Labeled sent message as Nurturing: {Subject}", subject);
            }
            else
            {
                _logger.LogDebug("No recent sent message found to label for subject: {Subject}", subject);
            }

            await client.DisconnectAsync(true, ct);
        }
        catch (Exception ex)
        {
            // Non-fatal — labeling failure should never break the send flow
            _logger.LogWarning(ex, "Failed to label sent message as Nurturing");
        }
    }

    // ── Helpers ─────────────────────────────────────────────────────

    /// <summary>
    /// Attaches both a plain-text and HTML version to the email as multipart/alternative.
    /// This improves deliverability (reduces spam score) because mail clients can pick
    /// the format they prefer and spam filters see a legitimate plain-text body.
    /// </summary>
    private static void AttachHtmlWithPlainTextAlternative(MailMessage msg, string htmlBody)
    {
        var plainText = HtmlToPlainText(htmlBody);

        // Set plain text as the body (fallback)
        msg.Body = plainText;
        msg.IsBodyHtml = false;

        // Add HTML as an alternate view — mail clients that support HTML will prefer this
        var htmlView = AlternateView.CreateAlternateViewFromString(htmlBody, System.Text.Encoding.UTF8, "text/html");
        msg.AlternateViews.Add(htmlView);
    }

    /// <summary>
    /// Converts HTML to readable plain text by stripping tags and decoding entities.
    /// Preserves paragraph breaks and link URLs.
    /// </summary>
    private static string HtmlToPlainText(string html)
    {
        if (string.IsNullOrWhiteSpace(html)) return "";

        var text = html;

        // Convert <br>, </p>, </div>, </tr> to newlines
        text = Regex.Replace(text, @"<br\s*/?>", "\n", RegexOptions.IgnoreCase);
        text = Regex.Replace(text, @"</(?:p|div|tr|li|h[1-6])>", "\n", RegexOptions.IgnoreCase);
        text = Regex.Replace(text, @"<hr[^>]*>", "\n---\n", RegexOptions.IgnoreCase);

        // Extract link URLs: <a href="URL">text</a> → text (URL)
        text = Regex.Replace(text, @"<a\s[^>]*href\s*=\s*[""']([^""']+)[""'][^>]*>(.*?)</a>",
            m =>
            {
                var url = m.Groups[1].Value;
                var linkText = m.Groups[2].Value;
                var cleanLinkText = Regex.Replace(linkText, @"<[^>]+>", "").Trim();
                return string.Equals(cleanLinkText, url, StringComparison.OrdinalIgnoreCase)
                    ? url
                    : $"{cleanLinkText} ({url})";
            },
            RegexOptions.IgnoreCase | RegexOptions.Singleline);

        // Strip all remaining HTML tags
        text = Regex.Replace(text, @"<[^>]+>", "");

        // Decode HTML entities
        text = WebUtility.HtmlDecode(text);

        // Collapse multiple blank lines into at most two
        text = Regex.Replace(text, @"\n{3,}", "\n\n");

        return text.Trim();
    }

    /// <summary>Creates a configured SMTP client for sending emails.</summary>
    private System.Net.Mail.SmtpClient CreateSmtpClient()
    {
        var smtp = new System.Net.Mail.SmtpClient(_options.SmtpHost!, _options.SmtpPort);
        smtp.EnableSsl = true;
        smtp.Credentials = new NetworkCredential(_options.EffectiveLoginEmail, _options.EmailPassword);
        return smtp;
    }
}

File: web-novit-ai/src/agent/community_manager/services/nurturing_mail_service.py
"""Nurturing mail service - IMAP/SMTP email operations.

Ported from NurturingMailService.cs.
"""

from __future__ import annotations

import email
import email.mime.multipart
import email.mime.text
import html
import imaplib
import logging
import re
import smtplib
from dataclasses import dataclass
from email.header import decode_header
from email.utils import parseaddr

from community_manager.config.settings import get_settings

logger = logging.getLogger(__name__)


@dataclass
class IncomingEmail:
    message_id: str
    from_address: str
    subject: str
    body: str


class NurturingMailService:
    """IMAP/SMTP email service for nurturing operations."""

    def __init__(self) -> None:
        self.settings = get_settings()

    @property
    def is_configured(self) -> bool:
        return bool(
            self.settings.nurturing_imap_host
            and self.settings.nurturing_smtp_host
            and self.settings.nurturing_login_email
            and self.settings.nurturing_email_password
        )

    async def get_unanswered_replies(self, days: int = 30) -> list[IncomingEmail]:
        """Get unanswered email replies from the last N days."""
        if not self.is_configured:
            logger.warning("Mail service not configured")
            return []

        try:
            import asyncio
            return await asyncio.to_thread(self._get_unanswered_replies_sync, days)
        except Exception:
            logger.exception("Failed to get unanswered replies")
            return []

    def _get_unanswered_replies_sync(self, days: int) -> list[IncomingEmail]:
        """Synchronous IMAP implementation."""
        imap = imaplib.IMAP4_SSL(
            self.settings.nurturing_imap_host,
            self.settings.nurturing_imap_port or 993,
        )
        try:
            imap.login(
                self.settings.nurturing_login_email,
                self.settings.nurturing_email_password,
            )

            # Get sent message IDs from Sent folder
            sent_message_ids = set()
            answered_message_ids = set()

            try:
                imap.select('"[Gmail]/Sent Mail"', readonly=True)
                status, messages = imap.search(None, "ALL")
                if status == "OK":
                    for msg_num in messages[0].split():
                        status, msg_data = imap.fetch(msg_num, "(BODY[HEADER.FIELDS (MESSAGE-ID)])")
                        if status == "OK" and msg_data[0]:
                            msg_id = self._extract_header_value(msg_data[0][1], "Message-ID")
                            if msg_id:
                                sent_message_ids.add(msg_id)

                        # Check for In-Reply-To (our replies)
                        status, msg_data = imap.fetch(msg_num, "(BODY[HEADER.FIELDS (IN-REPLY-TO)])")
                        if status == "OK" and msg_data[0]:
                            in_reply_to = self._extract_header_value(msg_data[0][1], "In-Reply-To")
                            if in_reply_to:
                                answered_message_ids.add(in_reply_to)
            except Exception:
                logger.warning("Could not read Sent folder", exc_info=True)

            # Search inbox for recent messages
            imap.select("INBOX", readonly=True)
            status, messages = imap.search(None, f'(SINCE "{self._days_ago(days)}")')

            if status != "OK":
                return []

            replies = []
            login_email = self.settings.nurturing_login_email.lower()
            sender_email = (self.settings.nurturing_sender_email or self.settings.nurturing_login_email).lower()

            for msg_num in messages[0].split():
                status, msg_data = imap.fetch(msg_num, "(RFC822)")
                if status != "OK":
                    continue

                msg = email.message_from_bytes(msg_data[0][1])
                from_addr = parseaddr(msg.get("From", ""))[1].lower()

                # Skip messages from ourselves
                if from_addr == login_email or from_addr == sender_email:
                    continue

                msg_id = msg.get("Message-ID", "")

                # Check if we already answered this
                if msg_id in answered_message_ids:
                    continue

                # Check if this is a reply to something we sent
                in_reply_to = msg.get("In-Reply-To", "")
                references = msg.get("References", "")
                is_reply_to_us = False

                for sent_id in sent_message_ids:
                    if sent_id in in_reply_to or sent_id in references:
                        is_reply_to_us = True
                        break

                if not is_reply_to_us:
                    continue

                # Extract body
                body = self._extract_body(msg)
                subject = self._decode_subject(msg.get("Subject", ""))

                replies.append(IncomingEmail(
                    message_id=msg_id,
                    from_address=from_addr,
                    subject=subject,
                    body=body,
                ))

            return replies

        finally:
            try:
                imap.logout()
            except Exception:
                pass

    async def send_email(
        self,
        to: str,
        subject: str,
        body: str,
        is_html: bool = False,
    ) -> None:
        """Send a single email."""
        if not self.is_configured:
            logger.warning("Mail service not configured")
            return

        import asyncio
        await asyncio.to_thread(self._send_email_sync, to, subject, body, is_html)

    def _send_email_sync(self, to: str, subject: str, body: str, is_html: bool) -> None:
        """Synchronous SMTP implementation."""
        sender_email = self.settings.nurturing_sender_email or self.settings.nurturing_login_email
        sender_name = self.settings.nurturing_sender_name or "Nicolas Piccardo"

        msg = email.mime.multipart.MIMEMultipart("alternative")
        msg["From"] = f"{sender_name} <{sender_email}>"
        msg["To"] = to
        msg["Subject"] = subject

        if is_html:
            # Add plain text alternative
            plain_text = self._html_to_plain_text(body)
            msg.attach(email.mime.text.MIMEText(plain_text, "plain", "utf-8"))
            msg.attach(email.mime.text.MIMEText(body, "html", "utf-8"))
        else:
            msg.attach(email.mime.text.MIMEText(body, "plain", "utf-8"))

        with smtplib.SMTP(self.settings.nurturing_smtp_host, self.settings.nurturing_smtp_port or 587) as server:
            server.starttls()
            server.login(self.settings.nurturing_login_email, self.settings.nurturing_email_password)
            server.send_message(msg)

    async def send_reply(
        self,
        to: str,
        subject: str,
        body: str,
        in_reply_to: str,
    ) -> None:
        """Send a reply to an existing email thread."""
        if not self.is_configured:
            return

        # Add Re: prefix if not already present
        if not subject.lower().startswith("re:"):
            subject = f"Re: {subject}"

        import asyncio
        await asyncio.to_thread(self._send_reply_sync, to, subject, body, in_reply_to)

    def _send_reply_sync(self, to: str, subject: str, body: str, in_reply_to: str) -> None:
        """Synchronous SMTP reply implementation."""
        sender_email = self.settings.nurturing_sender_email or self.settings.nurturing_login_email
        sender_name = self.settings.nurturing_sender_name or "Nicolas Piccardo"

        msg = email.mime.text.MIMEText(body, "plain", "utf-8")
        msg["From"] = f"{sender_name} <{sender_email}>"
        msg["To"] = to
        msg["Subject"] = subject
        msg["In-Reply-To"] = in_reply_to
        msg["References"] = in_reply_to

        with smtplib.SMTP(self.settings.nurturing_smtp_host, self.settings.nurturing_smtp_port or 587) as server:
            server.starttls()
            server.login(self.settings.nurturing_login_email, self.settings.nurturing_email_password)
            server.send_message(msg)

    async def mark_as_read(self, message_ids: list[str]) -> None:
        """Mark messages as read and move to Nurturing folder."""
        if not self.is_configured or not message_ids:
            return

        import asyncio
        await asyncio.to_thread(self._mark_as_read_sync, message_ids)

    def _mark_as_read_sync(self, message_ids: list[str]) -> None:
        """Synchronous IMAP mark as read."""
        imap = imaplib.IMAP4_SSL(
            self.settings.nurturing_imap_host,
            self.settings.nurturing_imap_port or 993,
        )
        try:
            imap.login(self.settings.nurturing_login_email, self.settings.nurturing_email_password)
            imap.select("INBOX")

            for msg_id in message_ids:
                status, messages = imap.search(None, f'(HEADER Message-ID "{msg_id}")')
                if status == "OK" and messages[0]:
                    for msg_num in messages[0].split():
                        imap.store(msg_num, "+FLAGS", "\\Seen")
                        # Try to move to Nurturing folder
                        try:
                            imap.copy(msg_num, "Nurturing")
                            imap.store(msg_num, "+FLAGS", "\\Deleted")
                        except Exception:
                            logger.warning("Could not move message to Nurturing folder")

            imap.expunge()
        finally:
            try:
                imap.logout()
            except Exception:
                pass

    def _extract_header_value(self, header_bytes: bytes, header_name: str) -> str | None:
        """Extract a header value from raw header bytes."""
        try:
            text = header_bytes.decode("utf-8", errors="replace")
            for line in text.split("\n"):
                if line.lower().startswith(header_name.lower() + ":"):
                    return line.split(":", 1)[1].strip().strip("<>")
        except Exception:
            pass
        return None

    def _extract_body(self, msg: email.message.Message) -> str:
        """Extract text body from an email message."""
        if msg.is_multipart():
            for part in msg.walk():
                content_type = part.get_content_type()
                if content_type == "text/plain":
                    payload = part.get_payload(decode=True)
                    if payload:
                        return payload.decode("utf-8", errors="replace")
        else:
            payload = msg.get_payload(decode=True)
            if payload:
                return payload.decode("utf-8", errors="replace")
        return ""

    def _decode_subject(self, subject: str) -> str:
        """Decode encoded email subject."""
        try:
            decoded_parts = decode_header(subject)
            parts = []
            for part, charset in decoded_parts:
                if isinstance(part, bytes):
                    parts.append(part.decode(charset or "utf-8", errors="replace"))
                else:
                    parts.append(part)
            return " ".join(parts)
        except Exception:
            return subject

    def _days_ago(self, days: int) -> str:
        """Get date string for N days ago (IMAP format)."""
        from datetime import datetime, timedelta
        d = datetime.now() - timedelta(days=days)
        return d.strftime("%d-%b-%Y")

    def _html_to_plain_text(self, html_body: str) -> str:
        """Convert HTML to plain text for multipart/alternative."""
        if not html_body:
            return ""

        text = html_body
        # Convert common HTML elements to plain text
        text = re.sub(r"<br\s*/?>", "\n", text, flags=re.IGNORECASE)
        text = re.sub(r"</(p|div|tr)>", "\n", text, flags=re.IGNORECASE)
        text = re.sub(r"<[^>]+>", " ", text)
        text = html.unescape(text)
        text = re.sub(r"\s+", " ", text)
        text = re.sub(r"\n\s*\n", "\n\n", text)
        return text.strip()
